parachain/src/primitives.rs

```diff
@@ -371,9 +382,10 @@ pub struct ValidationResult {
 	/// An update to the validation code that should be scheduled in the relay chain.
 	pub new_validation_code: Option<ValidationCode>,
 	/// Upward messages send by the Parachain.
-	pub upward_messages: Vec<UpwardMessage>,
+	pub upward_messages: BoundedVec<UpwardMessage, ConstU32<MAX_UPWARD_MESSAGE_NUM>>,
```

Comment:
Using a bounded vec doesn't solve the issue you are trying to solve here. https://github.com/paritytech/substrate/blob/2dff067e9f7f6f3cc4dbfdaaa97753eccc407689/primitives/core/src/bounded/bounded_vec.rs#L311-L315

If we look into this decode implementation, it rejects the vector after decoding it. You want to ensure that you don't even start try decoding/allocating when the length of the vector is more than the allowed maximum. Otherwise you still have the same issue.

We could probably change the substrate implementation to directly work this way. You first decode length of the vector and then early reject if that is too long.

Reply:
Ohh, I missed that - thanks!

I think changing the Substrate impl would be ideal. I can do a custom impl here, but I'd need to do it three times, because I found out during this PR that there are three separate `ValidationResult` types 🤷‍♂️.

Reply:
@bkchr Would this work? I can raise a PR in Substrate.

```rs
fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
    if let Some(len) = input.remaining_len()? {
        if len > S::get() as usize {
            return Err("BoundedVec exceeds its limit".into())
        }
    }
    let inner = Vec::<T>::decode(input)?;
    Ok(Self(inner, PhantomData))
}
```

Reply:
No, the following should work:
```
fn decode<I: codec::Input>(input: &mut I) -> Result<Self, codec::Error> {
    let len = Compact::<u32>::decode(input)?;

    if len > S::get() as usize {
        return Err("BoundedVec exceeds its limit".into())
    }
    
    let inner = decode_vec_with_length(len)?
    Ok(Self(inner, PhantomData))
}
```

We should probably make this function public to use it here: https://github.com/paritytech/parity-scale-codec/blob/4993588b96db80341e4a8008861e1db16c5831f6/src/codec.rs#L833-L836

Reply:
Okay, that would work too if you think it's a good idea. Should we also export the other functions like that, e.g. `decode_array`?

Reply:
No.